### PR TITLE
Fixed #30057 -- Fixed diffsettings leaving out custom default settings

### PR DESCRIPTION
**Pull request by orlnub123**
The changes I've made to the tests are experimental; if something could be improved I'd love your feedback.

### CODE REVIEW

**Comment by charettes** [comment]
@hramezani you might be interested in reviewing this patch since your worked on #10316 and have a bit of context.

**Comment by hramezani** [comment]
I just added the test for the [#10316](https://github.com/django/django/pull/10316) and the solution proposed in the [the ticket](https://code.djangoproject.com/ticket/29236) and also there was an unmerged pr [#9812](https://github.com/django/django/pull/9812) that based on the ticket solution.
BTW, Based on the added test for this patch(`test_dynamic_settings_configured`), it seems the problem does not related to [#10316](https://github.com/django/django/pull/10316) and by reverting the changes added by [#10316](https://github.com/django/django/pull/10316), the test doesn't pass and we will get `ImproperlyConfigured` error.
As @orlnub123 mentioned in [ticket_30057](https://code.djangoproject.com/ticket/30057), the reason is because `module_to_dict` reads the settings from the `__dict__` and `UserSettingsHolder` doesn't store settings from the default settings in its `__dict__`.

File: tests/admin_scripts/tests.py [hunk 1]
```diff
@@ -106,6 +106,25 @@ def remove_settings(self, filename, is_dir=False):
...
+
+        with open(destination, 'r') as fp:
+            contents = fp.read()
+        contents = contents.replace('{{ extra }}', extra)
+        contents = contents.replace('{{ args }}', args)
+        with open(destination, 'w') as fp:
+            fp.write(contents)
+
+    def remove_configured_settings(self):
+        path = os.path.join(
```
Comment (timgraham) [inline tests/admin_scripts/tests.py]: Use a single line. Our [style guide](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style) allows lines up to 119 characters if it helps readability.

File: tests/admin_scripts/tests.py [hunk 2]
```diff
@@ -166,20 +185,26 @@ def safe_remove(path):
             except OSError:
                 pass
 
+        configured_settings_file = os.path.join(
+            os.path.dirname(__file__), 'configured_settings_manage.py')
+        conf_dir = os.path.dirname(conf.__file__)
+        if configured_settings and not os.path.isfile(configured_settings_file):
+            self.write_configured_settings()
+            self.addCleanup(self.remove_configured_settings)
```
Comment (timgraham) [inline tests/admin_scripts/tests.py]: I'd think `write_configured_settings()` could be responsible for doing this.

File: django/conf/__init__.py [hunk 3]
```diff
@@ -180,6 +180,9 @@ def __init__(self, settings_module):
             os.environ['TZ'] = self.TIME_ZONE
             time.tzset()
 
+    def __dir__(self):
```
Comment (timgraham) [inline django/conf/__init__.py]: I reverted this change and didn't see any test failures.

File: tests/admin_scripts/tests.py [hunk 4]
```diff
@@ -2234,10 +2259,32 @@ def test_basic(self):
...
     def test_settings_configured(self):
+        self.write_configured_settings(args='DEBUG=True')
+        self.addCleanup(self.remove_configured_settings)
         out, err = self.run_manage(['diffsettings'], configured_settings=True)
         self.assertNoOutput(err)
         self.assertOutput(out, 'DEBUG = True')
 
+    def test_dynamic_settings_configured(self):
+        self.write_configured_settings(
+            extra='''class Settings:
```
Comment (timgraham) [inline tests/admin_scripts/tests.py]: The purpose of this class isn't obvious to me. Is this code adapted your use case? Maybe you can add a bit more explanation to the ticket. I'm not sure what "custom default settings" are.

File: tests/admin_scripts/tests.py [hunk 5]
```diff
@@ -166,20 +185,26 @@ def safe_remove(path):
...
-            manage_py_contents = fp.read()
-        manage_py_contents = manage_py_contents.replace(
-            "{{ project_name }}", "test_project")
-        with open(test_manage_py, 'w') as fp:
-            fp.write(manage_py_contents)
+        if not configured_settings:
+            with open(test_manage_py, 'r') as fp:
+                manage_py_contents = fp.read()
+            manage_py_contents = manage_py_contents.replace(
+                "{{ project_name }}", "test_project")
```
Comment (timgraham) [inline tests/admin_scripts/tests.py]: Use single quotes (unless a string contains a single quote) as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/admin_scripts/tests.py (hunk 4 above)
Comment (timgraham) [reply tests/admin_scripts/tests.py]: It's not obvious to me that the template approach is the best solution for readability as opposed to just creating another test settings file.

**Comment by collinanderson** [comment]
@hramezani would it work to do a simpler change, like this? I created a simpler diff based on your dir() method of module_to_dict. https://github.com/django/django/pull/10838

**Comment by timgraham** [comment]
I'll go with https://github.com/django/django/pull/10838. If it's not sufficient, please reopen the ticket.

**Comment by orlnub123** [comment]
Thanks for the reviews and the help. I'm glad the fix was applied in time for the alpha but I'm not sure if I agree with the implementation.

The additional isupper() check in the omittable serves the purpose of preventing the is_overridden method from showing up when querying regular unconfigured settings but I feel that it's a bit overkill. Lowercase settings that are otherwise normally accessible also end up getting omitted.
I also don't think the supplied test is adequate. It tests if you can set arbitrary settings which is nice to have but it doesn't address the core issue.

It seems GitHub ignores force-pushes if the PR is closed so I've made a new one: #11012.
Sorry for the inconvenience.

File: django/conf/__init__.py (hunk 3 above)
Comment (orlnub123) [reply django/conf/__init__.py]: I've added a test for it in #11012 but it'll only fail before 573f44d.

File: tests/admin_scripts/tests.py (hunk 4 above)
Comment (orlnub123) [reply tests/admin_scripts/tests.py]: It is somewhat. All that's needed is an object with an attribute that doesn't appear or has its value different from the default and is listed in the dir().

By default settings I meant the default_settings argument passed to settings.configure(). I tried to follow the terminology from the [settings docs](https://docs.djangoproject.com/en/dev/topics/settings/#custom-default-settings).

**Comment by timgraham** [comment]
[The documentation](https://docs.djangoproject.com/en/dev/topics/settings/#creating-your-own-settings) says, "Setting names must be all uppercase."

**Pull request closed without merging**